File: 219_Contains_Duplicate_II.py
# 2018/11/13 #1
class Solution:
	def containsNearbyDuplicate(self, nums, k):
		if not nums: return False
		dic={}
		for i,n in enumerate(nums):
			if n in dic and i-dic[n]<=k:
				return True
			else:
				dic[n]=i
		return False


slt=Solution()
inp=[1,2,3,1]
t=3
ans=True
print(slt.containsNearbyDuplicate(inp,t)==ans)
inp=[1,0,1,1]
t=1
ans=True
print(slt.containsNearbyDuplicate(inp,t)==ans)
inp=[1,2,3,1,2,3]
t=2
ans=False
print(slt.containsNearbyDuplicate(inp,t)==ans)

File: 219_Contains_Duplicate_II.go
// 2020/4/18
package main

import (
	"fmt"
	// "math"
	"time"
)

func main() {
	start := time.Now()
	nums := []int{1, 2, 3, 1}
	k := 3
	fmt.Println(containsNearbyDuplicate(nums, k))
	nums = []int{1, 0, 1, 1}
	k = 1
	fmt.Println(containsNearbyDuplicate(nums, k))
	nums = []int{1, 2, 3, 1, 2, 3}
	k = 2
	fmt.Println(!containsNearbyDuplicate(nums, k))
	elapsed := time.Now().Sub(start)
	fmt.Println(elapsed)
}

func containsNearbyDuplicate(nums []int, k int) bool {
	if len(nums) == 0 {
		return false
	}
	dic := make(map[int]int)
	for i, v := range nums {
		if _, ok := dic[v]; ok && i-dic[v] <= k {
			return true
		} else {
			dic[v] = i
		}
	}
	return false
}
